### PR TITLE
12747 [a11y] Make sure IDs are unique

### LINKED ISSUE #12747: 508-defect-1 : IDs of active elements MUST be unique
## Issue Description

There is a `IDs of active elements MUST be unique` accessibility error when multiple editable sections expanded on the `ReviewCollapsibleChapter`. 

It was identified on a specific app, but seems to be coming from platform code, not application code. 

### Screen shot

screen recording

### References

- Original Issue: https://github.com/department-of-veterans-affairs/va.gov-team/issues/12674
- Slack thread: https://dsva.slack.com/archives/CBU0KDSB1/p1598031421089900

## Front end tasks
- Ensure unit tests are available. 
- Ensure your code changes are covered by E2E tests
- Run axe checks using the Chrome or Firefox browser plugin

## Acceptance Criteria
- [ ] Unit tests pass
- [ ] E2E tests cover current code and regression
- [ ] End-to-end tests show 0 violations.
- [ ] All axe checks pass

## Appendix
<details> 

 <summary>Ensure your code changes are covered by E2E tests (expand)</summary>
 
 - Add E2E tests if none exist for this addition/change. 
 
 - Update existing E2E tests if this code will change functionality. 
 
 - Include axe checks, including hidden content
 
</details>

<details> 

 <summary> Run axe checks using the Chrome or Firefox browser plugin (expand)</summary>
 
- Ensure no heading levels are skipped.

- Ensure all buttons and labeled inputs use semantic HTML elements.

- Ensure all buttons, labeled elements and images are identified using HTML semantic markers or ARIA roles.

- Ensure form fields have clearly defined boundaries or outlines.

- Ensure form fields do not use placeholder text. 

- Ensure form fields have highly visible and specific error states. 

</details>

<details> 

 <summary> Test for color contrast and color blindness issues (expand) </summary>
 - All text has appropriate contrast. 

</details>

<details> 

 <summary> Zoom layouts to 400% at 1280px width (expand)</summary> 
 - Ensure readability and usability are supported when zoomed up to 400% at 1280px browser width
 - Ensure no content gets focused offscreen or is hidden from view.
 
</details>

<details> 

 <summary> Test with 1 or 2 screen readers (expand)</summary> 
 
 - Ensure the page includes a skip navigation link. 
 
 - Ensure all links are properly descriptive. 
 
 - Ensure screen reader users can hear the text equivalent for each image conveying information. 

 - Ensure screen reader users can hear the text equivalent for each image button. 

 - Ensure screen reader users can hear labels and instructions for inputs. 
 
 - Ensure purely decorative images are not announced by the screenreader. 
 
</details>

<details> 

 <summary>Navigate using the keyboard only (expand)</summary> 
 
 - Ensure all links (navigation, text and/or image), form controls and page functions can be reached with the tab key in a logical order. 
 
 - Ensure all links (navigation, text and/or image), form controls and page functions can be triggered with the spacebar, enter key, or arrow keys.
 
 - Ensure all interactive elements can be reached with the tab key in a logical order 
 
 - Ensure all interactive elements can be triggered with the spacebar, enter key, or arrow keys.
 
 - Ensure focus is always visible and appears in logical order.
 
 - Ensure each interactive element has visible focus state which appears in logical order.
 

</details>

---
## How to configure this issue
- [ ] **Attached to a Milestone** (when will this be completed?)
- [ ] **Attached to an Epic** (what body of work is this a part of?)
- [ ] **Labeled with Team** (`product support`, `analytics-insights`, `operations`, `service-design`, `tools-be`, `tools-fe`)
- [ ] **Labeled with Practice Area** (`backend`, `frontend`, `devops`, `design`, `research`, `product`, `ia`, `qa`, `analytics`, `contact center`, `research`, `accessibility`, `content`)
- [ ] **Labeled with Type** (`bug`, `request`, `discovery`, `documentation`, etc.)

### PR DESCRIPTION
**Pull request by bkjohnson**
## Description

Fixes https://github.com/department-of-veterans-affairs/va.gov-team/issues/12747

When forms allow multiple "items" to be added, there exists a problem where the `id` attribute for each added item would be the same, causing an accessibility issue for screen readers.

This PR changes it so that when successive items are added, their `id` in the DOM is appended with an integer corresponding to their order.

## Testing done

Local browser testing & unit testing

## Screenshots

Here you can see a `1` at the end of the id, which is there because the first item in the military service history list has a `0` appended to the end.  This pattern avoids the accessibility issue.

image

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/platform/forms-system/src/js/review/ArrayField.jsx [hunk 1]
```diff
@@ -223,9 +285,14 @@ class ArrayField extends React.Component {
             const showReviewButton =
               !itemCountLocked &&
               (!schema.minItems || items.length > schema.minItems);
-            const itemSchema = this.getItemSchema(index);
+            const itemSchema = getNumeratedItemSchema(this.props.schema, index);
```
Comment (bkjohnson) [inline src/platform/forms-system/src/js/review/ArrayField.jsx:288]: This ensures that the `schema` is numerated appropriately.

File: src/platform/forms-system/src/js/review/ArrayField.jsx [hunk 2]
```diff
@@ -223,9 +285,14 @@ class ArrayField extends React.Component {
...
             const itemTitle = itemSchema ? itemSchema.title : '';
 
+            const numeratedUiSchema = getNumeratedUiSchema(
+              uiSchema.items,
+              index,
+            );
```
Comment (bkjohnson) [inline src/platform/forms-system/src/js/review/ArrayField.jsx:294]: And this ensures that the `uiSchema` is numerated appropriately.

File: src/platform/forms-system/src/js/helpers.js [hunk 3]
```diff
@@ -642,3 +644,53 @@ export function transformForSubmit(
...
+ * Almost a copy of the class method `getItemSchema`.
+ * This version applies numeration to the `properties` and `required` values
+ *
+ * @param {Object} schema - The schema for the field
+ * @param {number} index - The integer which will be appended to the property names
+ * @returns {Object} The object which *might* have integers appended to property names
+ */
+export function getNumeratedItemSchema(schema, index) {
+  if (schema.items.length > index) {
+    return schema.items[index];
```
Comment (bkjohnson) [inline src/platform/forms-system/src/js/helpers.js:683]: I may need to enumerate things here as well :thinking:

File: src/platform/forms-system/test/js/helpers.unit.spec.js [hunk 4]
```diff
@@ -1003,4 +1005,98 @@ describe('Schemaform helpers:', () => {
...
+      firstName: {
+        'ui:title': 'First Name',
+      },
+      lastName: {
+        'ui:title': 'Last Name',
+      },
+    };
+
+    it('should numerate the keys for uiSchema fields', () => {
+      const numeratedUiSchema = getNumeratedUiSchema(uiSchema, 4);
```
Comment (billfienberg) [inline src/platform/forms-system/test/js/helpers.unit.spec.js:1024]: `4` is an arbitrary `index`, correct?

Comment (bkjohnson) [reply src/platform/forms-system/test/js/helpers.unit.spec.js:1024]: Yup, it's arbitrary. When this function gets called in a `map` callback though, each successive call will have an index incremented by 1.

File: src/platform/forms-system/src/js/helpers.js [hunk 5]
```diff
@@ -642,3 +644,50 @@ export function transformForSubmit(
 
   return JSON.stringify(withoutViewFields, replacer) || '{}';
 }
+
+/**
+ * Numerate the field names (keys) in a uiSchema object.
```
Comment (cvalarida) [inline src/platform/forms-system/src/js/helpers.js:649]: Probably worth calling out in the description here that it intentionally avoids adding numbers to keys which begin with `ui:`.

File: src/platform/forms-system/src/js/helpers.js [hunk 6]
```diff
@@ -642,3 +644,50 @@ export function transformForSubmit(
...
+ *
+ * @param {Object} uiSchema - The uiSchema for the field
+ * @param {number} index = The integer to append to the field names
```
Comment (cvalarida) [inline src/platform/forms-system/src/js/helpers.js:652]:
```suggestion
 * @param {number} index - The integer to append to the field names
```

File: src/platform/forms-system/src/js/helpers.js [hunk 7]
```diff
@@ -642,3 +644,50 @@ export function transformForSubmit(
...
+      delete numerated[key];
+      return numerated;
+    },
+    { ...uiSchema },
+  );
+  /* eslint-enable no-param-reassign */
+}
+
+/**
+ * Almost a copy of the class method `getItemSchema`.
```
Comment (cvalarida) [inline src/platform/forms-system/src/js/helpers.js:674]: Does this refer to the `ArrayField` class method which you're deleting?

**Review by cvalarida — COMMENTED** [review]
The only thing that has me possibly concerned about this approach (renaming properties in the schema & uiSchema) is how that may trickle down to the `idSchema` and be reflected in the final submission to the API.

Have you checked the form state in the redux store to make sure this doesn't cause any problems with form submission? :thinking: We might could try submitting a form and seeing if it throws any validation errors. At any rate, I just want to make sure the submitted form data matches the schema.

File: src/platform/forms-system/src/js/helpers.js (hunk 7 above)
Comment (bkjohnson) [reply src/platform/forms-system/src/js/helpers.js:674]: It does - that's some leftover copypasta.  I'll update accordingly.

**Comment by bkjohnson** [comment]
I'm changing approaches a bit - going to rely on changes made to the `react-jsonschema-form` dependency instead: https://github.com/department-of-veterans-affairs/react-jsonschema-form/pull/7

File: src/platform/forms-system/src/js/review/ArrayField.jsx [hunk 8]
```diff
@@ -226,6 +225,13 @@ class ArrayField extends React.Component {
             const itemSchema = this.getItemSchema(index);
             const itemTitle = itemSchema ? itemSchema.title : '';
 
+            const idSchema = toIdSchema(
+              itemSchema,
+              itemSchema?.$id,
+              this.props.schema.definitions,
+              index,
+            );
```
Comment (bkjohnson) [inline src/platform/forms-system/src/js/review/ArrayField.jsx:233]: This creates an `idSchema`, and the the `index` arg is used as a suffix that gets appended to each generated id.

Next, we pass this as a prop to `<SchemaForm>`, which passes it as a prop to `<Form>` in the `react-jsonschema-form` library.

**Review by cvalarida — APPROVED** [review]

**Pull request merged**